### PR TITLE
Upgrading to v2.4.0

### PR DESCRIPTION
**Pull request by metinavci**
SCLJ-9 Reporting Battery Level Changing
SCLJ-10 Reporting Terminal Status
SCLJ-11 Free form printing
SCLJ-20 Reporting Terminal Configuration
SCLJ-21 Ability to suppress Merchant Password from POS for refund
SCLJ-22 Allow surcharging from POS
SCLJ-23 Automated Address Resolution - Implement device lookup for IP address using Serial Number
SCLJ-27 Java - The app becomes unresponsive when the user does not use it for a while (let it "idle")
SCLJ-29 The Device Address Updated keeps popping up when Unpairing is initiated from the terminal
SCLJ-33 Unable to unpair after recovering from network disconnect
SCLJ-34 Exception in RamenPOS when attempting to save with an invalid serial number
SCLJ-39 Reduce Timeout and Re-connection Duration

### CODE REVIEW

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 1]
```diff
@@ -117,6 +132,13 @@ public SpiPayAtTable enablePayAtTable() {
         return spiPat;
     }
 
+    public SpiPayAtTable disablePayAtTable()
+    {
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: Formatting.

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 2]
```diff
@@ -156,6 +178,72 @@ public void start() {
         }
     }
 
+    /**
+     * Set the acquirer code of your bank, please contact Assembly's Integration Engineers for acquirer code.
+     */
+    public boolean setAcquirerCode(String acquirerCode) {
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: What does the return value represent if it's always true?

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 3]
```diff
@@ -156,6 +178,72 @@ public void start() {
...
+        this.deviceApiKey = deviceApiKey;
+        return true;
+    }
+
+    /**
+     * Allows you to set the serial number of the Eftpos
+     */
+    public boolean setSerialNumber(String serialNumber) throws Exception {
+        if (getCurrentStatus() != SpiStatus.UNPAIRED)
+            return false;
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: Formatting. If statements with no braces on two lines are discouraged. For `return false;` it's fine to just move it to the same line as the if-statement, for more complex statements (like the one a few lines down), add curly braces.

This applies to all instances in this PR, so I won't leave this comment on all of them.

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 4]
```diff
@@ -168,12 +256,12 @@ public boolean setPosId(@NotNull String id) {
...
 
     /**
-     * Allows you to set the PIN pad address. Sometimes the PIN pad might change IP address (we recommend
+     * Allows you to set the PIN pad address only if auto address is not enabled. Sometimes the PIN pad might change IP address (we recommend
      * reserving static IPs if possible). Either way you need to allow your User to enter the IP address
      * of the PIN pad.
      */
     public boolean setEftposAddress(String address) {
-        if (getCurrentStatus() == SpiStatus.PAIRED_CONNECTED) return false;
+        if ((getCurrentStatus() == SpiStatus.PAIRED_CONNECTED) || autoAddressResolutionEnabled) return false;
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: Redundant parentheses.

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 5]
```diff
@@ -1297,6 +1567,20 @@ private void onSpiConnectionStatusChanged(@NotNull Connection.State state) {
...
+                            if (autoAddressResolutionEnabled) {
+                                if (retriesSinceLastDeviceAddressResolution >= retriesBeforeResolvingDeviceAddress) {
+                                    try {
+                                        autoResolveEftposAddress();
+                                    } catch (Exception e) {
+                                        LOG.error("Failed to connect to Auto IP Resolution Service", e);
+                                    }
+                                    retriesSinceLastDeviceAddressResolution = 0;
+                                } else {
+                                    retriesSinceLastDeviceAddressResolution += 1;
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: `retriesSinceLastDeviceAddressResolution++`

File: client/src/main/java/com/assemblypayments/spi/util/DeviceService.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger("spi");
+
+    public DeviceAddressStatus retrieveService(String serialNumber, String apiKey, String acquirerCode, boolean isTestMode) {
+        String deviceAddressUrl;
+        StringBuilder response;
+
+        try {
+            if (isTestMode) {
+                deviceAddressUrl = "https://device-address-api-sb." + acquirerCode + ".msp.assemblypayments.com/v1/" + serialNumber + "/ip";
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/util/DeviceService.java]: This is prone to typos, use 

```deviceAddressUrl = String.format("https://device-address-api%s.%s.msp.assemblypayments.com/v1/%s/ip", envSuffix, acquirerCode, serialNumber);```

and switch only the `envSuffix` between `""` and `"-sb"` based on `isTestMode`.

File: client/src/main/java/com/assemblypayments/spi/util/DeviceService.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            } else {
+                deviceAddressUrl = "https://device-address-api." + acquirerCode + ".msp.assemblypayments.com/v1/" + serialNumber + "/ip";
+            }
+
+            URL obj = new URL(deviceAddressUrl);
+
+            HttpURLConnection con = (HttpURLConnection) obj.openConnection();
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/util/DeviceService.java]: This method of opening HTTP connections is not very robust and you'll find it difficult to debug issues. You're better off just resolving issues with the OkHttp branch first and then rewriting this method as a single call with (see [OkHttp examples](http://square.github.io/okhttp/#examples)).

File: client/src/main/java/com/assemblypayments/spi/Spi.java (hunk 1 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/Spi.java]: Done

File: client/src/main/java/com/assemblypayments/spi/Spi.java (hunk 2 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/Spi.java]: you're right. But, I converted it from .Net Library which is developed by @RaymondK-AP. I change it to void method.

File: client/src/main/java/com/assemblypayments/spi/util/DeviceService.java (hunk 6 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/util/DeviceService.java]: Done

File: client/src/main/java/com/assemblypayments/spi/util/DeviceService.java (hunk 7 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/util/DeviceService.java]: it's done

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 8]
```diff
@@ -1270,6 +1535,9 @@ private void onSpiConnectionStatusChanged(@NotNull Connection.State state) {
                     LOG.info("I'm connected to " + eftposAddress + "...");
                     spiMessageStamp.setSecrets(secrets);
                     startPeriodicPing();
+
+                    // Clean up timer
+                    cleanReconnectFuture();
```
Comment (metinavci) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: @ellulpatrick I told the issue before. The reconnection thread triggers the DeviceAddressChanged delegation nonstop. When the library tries to reconnect this issue occurs.

onSpiConnectionStatusChangedin Spi.java

ScheduledThreadPoolExecutor runs nonstop after reconnecting in the DISCONNECTED case

This cleanup timer stops the thread.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.io.IOException;
+
+public class DeviceService {
+
+    private static final Logger LOG = LoggerFactory.getLogger("spi");
+
+    public DeviceAddressStatus retrieveService(String serialNumber, String apiKey, String acquirerCode, boolean isTestMode) {
+        String deviceAddressUrl;
+        OkHttpClient client = new OkHttpClient();
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: Client should be reused.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+        try {
+            String envSuffix = "";
+
+            if (isTestMode) {
+                envSuffix = "-sb";
+            }
+
+            deviceAddressUrl = String.format("https://device-address-api%s.%s.msp.assemblypayments.com/v1/%s/ip", envSuffix, acquirerCode, serialNumber);
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: Expressions that don't throw exceptions shouldn't be inside the try/catch clause.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+            String apiKeyHeader = "ASM-MSP-DEVICE-ADDRESS-API-KEY";
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: Redundant variable. It should either be a constant or just literal in the `addHeader()` expression.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+            Request request = new Request.Builder()
+                    .url(deviceAddressUrl)
+                    .addHeader(apiKeyHeader, apiKey)
+                    .build();
+
+            try (Response response = client.newCall(request).execute()) {
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: Redundant try within a try. Just write them out normally, you're already covered by the outside try.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+                return new Gson().fromJson(response.body() != null ? response.body().string() : null, DeviceAddressStatus.class);
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: * Gson instance should ideally be reused
* Unreadable line: I'm not against ternary operators, but at least store its output in a named variable

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 14]
```diff
@@ -156,6 +179,66 @@ public void start() {
...
+    /**
+     * Allows you to set the serial number of the Eftpos
+     */
+    public boolean setSerialNumber(String serialNumber) {
+        if (getCurrentStatus() != SpiStatus.UNPAIRED) return false;
+
+        String was = this.serialNumber;
+        this.serialNumber = serialNumber;
+        if (autoAddressResolutionEnabled && hasSerialNumberChanged(was))
+            autoResolveEftposAddress();
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: Braces please.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java (hunk 9 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: it's done

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java (hunk 10 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: it's done

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java (hunk 11 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: it's done

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java (hunk 12 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: it's done

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        try {
+            Request request = new Request.Builder()
+                    .url(deviceAddressUrl)
+                    .addHeader("ASM-MSP-DEVICE-ADDRESS-API-KEY", apiKey)
+                    .build();
+
+            Response response = okHttpClient.newCall(request).execute();
+
+            if (response.body() != null) {
+                deviceAddressStatus = new Gson().fromJson(response.body().string(), DeviceAddressStatus.class);
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: Gson still not reused.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+import java.io.IOException;
+
+public class DeviceService {
+
+    private static final Logger LOG = LoggerFactory.getLogger("spi");
+
+    public DeviceAddressStatus retrieveService(String serialNumber, String apiKey, String acquirerCode, boolean isTestMode) {
+        String envSuffix = "";
+        String deviceAddressUrl;
+        OkHttpClient okHttpClient = new OkHttpClient();
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: OkHttpClient still not reused.

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java (hunk 15 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: But they are only used in one method. Should it be still reusable?

File: client/src/main/java/com/assemblypayments/spi/service/DeviceService.java (hunk 16 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/service/DeviceService.java]: But it's only used in one method. Should it be still reusable?

**Pull request merged**